lib/krb5/pkinit-ec.c

```diff
@@ -172,51 +169,56 @@ _krb5_pk_rd_pa_reply_ecdh_compute_key(krb5_context context,
-                               N_("PKINIT: Can't parse ECDH public key", ""));
-        return ret;
-    }
-
-    *out_sz = (EC_GROUP_get_degree(group) + 7) / 8;
-    if (*out_sz < 0)
-        return EOVERFLOW;
-    *out = malloc(*out_sz);
-    if (*out == NULL) {
-        EC_KEY_free(public);
-        return krb5_enomem(context);
-    }
-    dh_gen_keylen = ECDH_compute_key(*out, *out_sz,
-                                     EC_KEY_get0_public_key(public),
-                                     ctx->u.eckey, NULL);
-    EC_KEY_free(public);
-    if (dh_gen_keylen <= 0) {
-        ret = KRB5KRB_ERR_GENERIC;
-        dh_gen_keylen = 0;
-        krb5_set_error_message(context, ret,
-                               N_("PKINIT: Can't compute ECDH public key", ""));
-        free(*out);
-        *out = NULL;
-        *out_sz = 0;
-    }
-    *out_sz = dh_gen_keylen;
+    EVP_PKEY_CTX *pctx = NULL;
+    EVP_PKEY *template = NULL;
+    EVP_PKEY *public = NULL;
+    size_t shared_len = 0;
+
+    if ((template = EVP_PKEY_new()) == NULL)
+        ret = krb5_enomem(context);
+    if (ret == 0 &&
+        EVP_PKEY_copy_parameters(template, ctx->u.eckey) == 0)
+        ret = krb5_enomem(context);
+    if (ret == 0 && (pctx = EVP_PKEY_CTX_new(ctx->u.eckey, NULL)) == NULL)
+        ret = krb5_enomem(context);
+    if (ret == 0 && EVP_PKEY_derive_init(pctx) != 1)
+        ret = krb5_enomem(context);
+    if (ret == 0 &&
+        EVP_PKEY_CTX_set_ecdh_kdf_type(pctx, EVP_PKEY_ECDH_KDF_NONE) != 1)
+        ret = krb5_enomem(context);
+    if (ret == 0 &&
+        (public = d2i_PublicKey(EVP_PKEY_EC, &template, &in, in_sz)) == NULL)
+        krb5_set_error_message(context,
+                               ret = HX509_PARSING_KEY_FAILED,
+                               "PKINIT: Can't parse the KDC's ECDH public key");
+    if (ret == 0 &&
+        EVP_PKEY_derive_set_peer_ex(pctx, public, 1) == 0)
+        krb5_set_error_message(context,
+                               ret = KRB5KRB_ERR_GENERIC,
+                               "Could not derive ECDH shared secret for PKINIT key exchange "
+                               "(EVP_PKEY_derive_set_peer_ex)");
+    if (ret == 0 &&
+        (EVP_PKEY_derive(pctx, NULL, &shared_len) == 0 || shared_len == 0))
+        krb5_set_error_message(context,
+                               ret = KRB5KRB_ERR_GENERIC,
+                               "Could not derive ECDH shared secret for PKINIT key exchange "
+                               "(EVP_PKEY_derive to get length)");
+    if (ret == 0 && shared_len > INT_MAX)
+        krb5_set_error_message(context,
+                               ret = KRB5KRB_ERR_GENERIC,
+                               "Could not derive ECDH shared secret for PKINIT key exchange "
+                               "(shared key too large)");
+    if (ret == 0 && (*out = malloc(shared_len)) == NULL)
+        ret = krb5_enomem(context);
+    if (ret == 0 && EVP_PKEY_derive(pctx, *out, &shared_len) == 0)
```

Comment:
I believe EVP_PKEY_derive() returns 0 upon success.

Reply:
![image](https://user-images.githubusercontent.com/604851/202961976-36b45d46-8ca7-49f1-b8fe-4afe26e46830.png)

Reply:
> ![image](https://user-images.githubusercontent.com/604851/202961976-36b45d46-8ca7-49f1-b8fe-4afe26e46830.png)

Upon re-reading the OpenSSL 3.0 source, it is consistent.   The call usage in the OpenSSL sources is "EVP_PKEY_derive() <= 0" return error.